daemon/cmd/daemon_main.go

```diff
@@ -1539,30 +1513,104 @@ func (d *Daemon) initKVStore() {
+		} else {
+			localNode.Update(func(n *nodeTypes.Node) {
+				n.SetNodeInternalIP(ip)
+			})
+		}
+	}
+
+	datapathConfig := linuxdatapath.DatapathConfiguration{
+		HostDevice: defaults.HostDevice,
+		ProcFs:     option.Config.ProcFs,
+	}
+
+	iptablesManager := &iptables.IptablesManager{}
+
+	lc.Append(hive.Hook{
+		OnStart: func(context.Context) error {
+			if err := enableIPForwarding(); err != nil {
+				log.Fatalf("enabling IP forwarding via sysctl failed: %s", err)
+			}
+
+			iptablesManager.Init()
+			return nil
+		}})
+
+	var wgAgent *wireguard.Agent
+	if option.Config.EnableWireguard {
+		switch {
+		case option.Config.EnableIPSec:
+			log.Fatalf("Wireguard (--%s) cannot be used with IPSec (--%s)",
+				option.EnableWireguard, option.EnableIPSecName)
+		case option.Config.EnableL7Proxy:
+			log.Fatalf("Wireguard (--%s) is not compatible with L7 proxy (--%s)",
+				option.EnableWireguard, option.EnableL7Proxy)
+		}
+
+		var err error
+		privateKeyPath := filepath.Join(option.Config.StateDir, wireguardTypes.PrivKeyFilename)
+		wgAgent, err = wireguard.NewAgent(privateKeyPath)
+		if err != nil {
+			log.Fatalf("failed to initialize wireguard: %s", err)
+		}
+
+		lc.Append(hive.Hook{
+			OnStop: func(context.Context) error {
+				wgAgent.Close()
+				return nil
+			},
+		})
+	} else {
+		// Delete wireguard device from previous run (if such exists)
+		link.DeleteByName(wireguardTypes.IfaceName)
+	}
+	return linuxdatapath.NewDatapath(datapathConfig, iptablesManager, wgAgent)
+}
+
+// daemonCell wraps the existing implementation of the cilium-agent that has
+// not yet been converted into a cell. Provides *Daemon as a Promise that is
+// resolved once daemon has been started to facilitate conversion into modules.
+var daemonCell = cell.Module(
+	"Daemon",
+
+	cell.Provide(newDaemonPromise),
+	cell.Invoke(func(promise.Promise[*Daemon]) {}), // Force instantiation.
+)
+
 type daemonParams struct {
 	cell.In
 
 	Lifecycle      hive.Lifecycle
-	Shutdowner     hive.Shutdowner
-	Config         DaemonCellConfig
-	LocalNodeStore node.LocalNodeStore
 	Clientset      k8sClient.Clientset
+	Datapath       datapath.Datapath
+	LocalNodeStore node.LocalNodeStore
+	Shutdowner     hive.Shutdowner
 }
 
-// registerDaemonHooks registers the lifecycle hooks for the part of the cilium-agent that has
-// not yet been modularized.
-//
-// The Daemon struct and objects referred to from there are not supplied to the graph as
-// object creation in Daemon is intertwined with side-effectful initialization. This stops
-// us from constructing and inspecting the dependency graph, so instead the daemon is constructed
-// and started via an OnStart hook.
-//
-// If an object still owned by Daemon is required in a module, it should be provided indirectly, e.g. via
-// a callback.
-func registerDaemonHooks(params daemonParams) error {
-	if params.Config.SkipDaemon {
-		return nil
-	}
+func newDaemonPromise(params daemonParams) promise.Promise[*Daemon] {
```

Comment:
Second day thoughts on the above is that most dependencies should already be handled by the Hive / Cell design & parameters being automatically resolved. But probably there is more use for promises in cases where a module runs its own dedicated goroutines for module-internal logic, and that logic needs to rely on the availability of some other module. I may have a good use for this in ipcache, since there's currently some ugly logic around doing some initial setup and then retrying the main loop if it gets triggered before identity allocation or k8s is fully synced. Not sure how much of that can be resolved via Cells vs. promises yet, but we can figure that out over time.

Reply:
The main reason for the `Promise[*Daemon]` right now is that we're using it in tests (daemon_test.go, controlplane) to grab `*Daemon` and use it to validate results.

We can also use it to "lift" things out from `*Daemon` and depend on them in other components, but of course preferably we would not depend on `Promise[*Foo]` and would rather provide `*Foo` normally via injection. I'll reword the commits to not suggest this and just document the current use.